core/src/cluster_info.rs

```diff
@@ -913,27 +913,37 @@ impl ClusterInfo {
     /// each Vec is no larger than `PROTOCOL_PAYLOAD_SIZE`
     /// Note: some messages cannot be contained within that size so in the worst case this returns
     /// N nested Vecs with 1 item each.
-    fn split_gossip_messages(mut msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
+    fn split_gossip_messages(msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
         let mut messages = vec![];
-        while !msgs.is_empty() {
-            let mut payload = vec![];
-            let mut size = serialized_size(&payload).expect("Couldn't check size");
-            while let Some(msg) = msgs.pop() {
-                let msg_size = msg.size();
-                if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                    if msg_size < MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                        msgs.push(msg);
-                    } else {
-                        debug!(
-                            "dropping message larger than the maximum payload size {:?}",
-                            msg
-                        );
+        let mut payload = vec![];
+        let base_size = serialized_size(&payload).expect("Couldn't check size");
+        let max_payload_size = MAX_PROTOCOL_PAYLOAD_SIZE - base_size;
+        let mut payload_size = 0;
+        for msg in msgs {
+            let msg_size = msg.size();
+            // If the message is too big to fit in this batch
+            if payload_size + msg_size > max_payload_size as u64 {
+                // See if it can fit in the next batch
+                if msg_size <= max_payload_size as u64 {
+                    if !payload.is_empty() {
+                        // Flush the  current payload
+                        messages.push(payload);
+                        // Init the next payload
+                        payload = vec![msg];
```

Comment:
it looks like msg is encoded twice, here and after it falls through to the end of the loop.  get some sleep and write better unit tests :)

Reply:
lol but it "continues" at the end of the if block?

Reply:
@aeyakovenko did you see the `continue` at the end? it's not encoding anything twice, its just resetting state, adding the first new item and going next.